internal/pkg/scaffold/olm-catalog/descriptor/parse.go

```diff
@@ -0,0 +1,261 @@
+		}
+	case 2:
+		switch pathElems[1] {
+		case "displayName":
+			d.DisplayName, err = strconv.Unquote(val)
+			if err != nil {
+				return errors.Wrapf(err, "error unquoting %s", val)
+			}
+		case "x-descriptors":
+			xdStr, err := strconv.Unquote(val)
+			if err != nil {
+				return errors.Wrapf(err, "error unquoting %s", val)
+			}
+			d.XDescriptors = strings.Split(xdStr, ",")
+		default:
+			return errors.Errorf("unsupported descriptor path element %s", pathElems[1])
+		}
+	default:
+		return errors.Errorf("unsupported descriptor path %s", annotations.JoinPath(pathElems...))
+	}
+	return nil
+}
+
+// parseResource parses a resource string of the form:
+// "kind,version,\"quoted name\""
+func parseResource(rStr string) (r olmapiv1alpha1.APIResourceReference, err error) {
+	rStr, err = strconv.Unquote(rStr)
+	if err != nil {
+		return r, err
+	}
+	rSplit := strings.SplitN(rStr, ",", 3)
+	if len(rSplit) < 2 {
+		return r, errors.Errorf("resource string %s did not have at least a kind and a version", rStr)
+	}
+	r.Kind, r.Version = strings.TrimSpace(rSplit[0]), strings.TrimSpace(rSplit[1])
+	if len(rSplit) == 3 {
+		r.Name, err = strconv.Unquote(rSplit[2])
+		if err != nil {
+			return r, err
+		}
+		r.Name = strings.TrimSpace(r.Name)
+	}
+	return r, nil
+}
+
+// parseDescription joins comment strings into one line, removing any tool
+// directives.
+func parseDescription(comments []string) string {
+	var lines []string
+	for _, c := range comments {
+		l := strings.TrimSpace(strings.TrimLeft(c, "/"))
+		if l == "" || strings.HasPrefix(l, "+") {
+			continue
+		}
+		lines = append(lines, l)
+	}
+	return strings.Join(lines, " ")
+}
+
+var jsonTagRe = regexp.MustCompile(`json:"([a-zA-Z0-9,]+)"`)
+
+func parsePathFromJSONTags(tags string) string {
+	tagMatches := jsonTagRe.FindStringSubmatch(tags)
+	if len(tagMatches) > 1 {
+		ts := strings.Split(tagMatches[1], ",")
+		if len(ts) != 0 && ts[0] != "" {
+			return ts[0]
+		}
+	}
+	return ""
+}
+
+// From https://github.com/openshift/console/blob/master/frontend/public/components/operator-lifecycle-manager/descriptors/types.ts#L5-L14
+var specXDescriptors = map[string]string{
+	"size":                 "urn:alm:descriptor:com.tectonic.ui:podCount",
+	"podCount":             "urn:alm:descriptor:com.tectonic.ui:podCount",
+	"endpoints":            "urn:alm:descriptor:com.tectonic.ui:endpointList",
+	"endpointList":         "urn:alm:descriptor:com.tectonic.ui:endpointList",
+	"label":                "urn:alm:descriptor:com.tectonic.ui:label",
+	"resources":            "urn:alm:descriptor:com.tectonic.ui:resourceRequirements",
+	"resourceRequirements": "urn:alm:descriptor:com.tectonic.ui:resourceRequirements",
+	"selector":             "urn:alm:descriptor:com.tectonic.ui:selector:",
+	"namespaceSelector":    "urn:alm:descriptor:com.tectonic.ui:namespaceSelector",
+	"booleanSwitch":        "urn:alm:descriptor:com.tectonic.ui:booleanSwitch",
```

Comment:
Wdyt about all constant annotations be an annotations.go file to make it clear?

Reply:
These are x-descriptors, not annotations. Their strings are only relevant when parsing. These change often, so having them be constantly-typed with names is unnecessary code. Having them in a lookup table is much easier.